base/compiler/abstractinterpretation.jl

```diff
@@ -1015,6 +1017,34 @@ function abstract_call_known(interp::AbstractInterpreter, @nospecialize(f),
     return abstract_call_gf_by_type(interp, f, argtypes, atype, sv, max_methods)
 end
 
+function abstract_call_opaque_closure(interp::AbstractInterpreter, clos::PartialOpaque, argtypes::Vector{Any}, sv::InferenceState)
+    if isa(clos.ci, CodeInfo)
+        nargtypes = argtypes[2:end]
+        pushfirst!(nargtypes, clos.env)
+        result = InferenceResult(Core.OpaqueClosure, nargtypes)
+        state = InferenceState(result, copy(clos.ci), false, interp)
+        typeinf_local(interp, state)
```

Comment:
This code effectively claims that an OpaqueClosure cannot call itself. However, that would seem to imply it cannot be passed as an argument, as otherwise we can build a y-combinator with it and invalidate this claim. It also implies that it can't call any recursive function that could call it (also a faulty assumption, for the same reason). I think that implies we need the majority of the same logic as abstract_call_method here. Probably by refactoring some helper functions?

Since we don't optimize much yet anyways, I'd suggested that currently it should treat all opaque objects as potential derivatives of the same template.